ElectronClient/plugins/GotoAnything.jsx

```diff
@@ -165,9 +186,55 @@ class Dialog extends React.PureComponent {
 					const path = Folder.folderPathString(this.props.folders, row.parent_id);
 					results[i] = Object.assign({}, row, { path: path ? path : '/' });
 				}
-			} else { // NOTES
+			} else if (this.state.query.indexOf('/') === 0) { // BODY
 				listType = BaseModel.TYPE_NOTE;
-				searchQuery = this.makeSearchQuery(this.state.query);
+				searchQuery = this.makeBodySearchQuery(this.state.query);
+				results = await SearchEngine.instance().search(searchQuery);
+
+				const limit = 20;
+				const searchKeywords = this.keywords(searchQuery);
+				const notes = await Note.byIds(results.map(result => result.id).slice(0, limit), { fields: ['id', 'body'] });
+				const notesById = notes.reduce((obj, { id, body }) => ((obj[[id]] = body), obj), {});
+
+				for (let i = 0; i < results.length; i++) {
+					const row = results[i];
+					let fragments = '...';
+
+					if (i < limit) {
+						const body = notesById[row.id];
+
+						const indices = [];
+						for (const { valueRegex } of searchKeywords) {
+							for (const match of body.matchAll(new RegExp(valueRegex, 'ig'))) {
+								indices.push(closestWordIndices(body, match, 15));
+								if (indices.length > 20) break;
+							}
+						}
+
+						// merge overlapping indices
+						indices.sort((a, b) => a[0] - b[0]);
+						const stack = [];
+						stack.push(indices[0]);
+						for (let j = 1; j < indices.length && stack.length < 3; j++) {
+							const top = stack[stack.length - 1];
+							if (top[1] < indices[j][0]) {
+								stack.push(indices[j]);
+							} else if (top[1] < indices[j][1]) {
+								top[1] = indices[j][1];
+								stack.pop();
+								stack.push(top);
+							}
+						}
+
+						fragments = stack.map(f => body.slice(f[0], f[1])).join(' ... ');
+					}
```

Comment:
Could you refactor this into a function to make it clearer what it does?

Reply:
I've refactored the merge code into a function and added comments to make it more understandable.